### PR TITLE
Suppport executing wasm workloads with wasmer

### PR DESCRIPTION
**Pull request by Furisto**
See [docs](https://github.com/containers/youki/blob/47da1927c3b61977907b1c198a5a29baca495a08/docs/src/user/webassembly.md)

### CODE REVIEW

**Comment by tsturzl** [comment]
This is really interesting. I'm wondering how you plan to distinguish between a container and a wasm workload, and how a high level runtime might integrate with this. This could be a really interesting alternative to something like Krustlet which could allow you to run wasm and container workloads on the same kubelet, whereas krustlet only supports wasm workloads on the hosts running it.

**Comment by Furisto** [comment]
@tsturzl This is based on the [compat](https://github.com/solo-io/wasm/blob/master/spec/spec-compat.md) variant of the Wasm Image Specification. The config.json needs to have an annotation of either run.oci.handler or module.wasm.image/variant in order for us to detect that a wasm workload has be run. The linked spec also describes how to build an image that is compliant with the Wasm Image spec. Here is an [example](https://github.com/second-state/wasmedge-containers-examples/blob/main/crio/README.md) of how you can use it with CRI-O.

**Comment by tsturzl** [comment]
I will also review in addition to @utam0k. This is a pretty big change so it's probably good to have a few eyes on it. This is awesome work though.

File: crates/libcontainer/src/workload/mod.rs [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+}
+pub struct CompositeExecutor {
+    executors: Vec<Box<dyn Executor>>,
+}
+
+impl Executor for CompositeExecutor {
+    fn exec(&self, spec: &Spec) -> Result<()> {
+        for executor in &self.executors {
+            if executor
+                .can_handle(spec)
```
[the remark this answers is not in the document]
Comment (tsturzl) [reply crates/libcontainer/src/workload/mod.rs:30]: And of course, I'm not demanding the change, but perhaps it's worth the conversation. Maybe there's things I'm not considering or overlooking here, or maybe others find the dynamic dispatch approach to be nicer also.

File: crates/libcontainer/src/workload/mod.rs [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+    /// Checks if the handler is able to handle the workload
+    fn can_handle(&self, spec: &Spec) -> Result<bool>;
+    /// The name of the handler
+    fn name(&self) -> &str;
+}
+pub struct CompositeExecutor {
+    executors: Vec<Box<dyn Executor>>,
+}
+
+impl Executor for CompositeExecutor {
```
Comment (tsturzl) [inline crates/libcontainer/src/workload/mod.rs:26]: What's the benefit for having a CompositeExecutor? It seems the major benefit for an Executor is that is creates a common interface for Executors, could this maybe just be a ExecutorManager that doesn't need to satisfy `Executor`? I don't imagine we'll ever need the `name` function here, and that's just dead code.

File: crates/libcontainer/src/workload/wasmer.rs [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+            .get_function("_start")
+            .context("could not retrieve wasm module main function")?;
+        start
+            .call(&[])
+            .context("wasm module was not executed successfuly")?;
+
+        Ok(())
+    }
+
+    fn can_handle(&self, spec: &Spec) -> Result<bool> {
```
Comment (tsturzl) [inline crates/libcontainer/src/workload/wasmer.rs:62]: In all the instances of `can_handle` Result isn't ever really needed, and always returns `Ok`. This seems like unnecessary indirection.

File: crates/libcontainer/src/workload/mod.rs [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+
+pub mod default;
+#[cfg(feature = "wasm-wasmer")]
+pub mod wasmer;
+
+static EMPTY: Vec<String> = Vec::new();
+
+pub trait Executor {
+    /// Executes the workload
+    fn exec(&self, spec: &Spec) -> Result<()>;
```
Comment (tsturzl) [inline crates/libcontainer/src/workload/mod.rs:16]: Do any of the functions for Executor actually need a reference to `self`? The CompositeExecutor is the only one that uses it, and it probably doesn't even need to satisfy this trait. If none of the Executors needed a `&self`, you'd also avoid the need to ever instantiate Executors, which aside from the CompositeExecutor are all empty structs.

**Review by tsturzl — CHANGES_REQUESTED** [review]
First of all, this is really cool stuff, and it's awesome that you took the time and initiative to do this work. So thank you for that. I'm not entirely on board with using dynamic dispatch here. I think perhaps I'd consider refactoring to get rid of the CompositeExecutor and opt for a ExecutorManager that doesn't need to satisfy the Executor trait. This ExecutorManager would instead statically dispatch hard coded Executors with a match block, and therefore only instantiate and allocate the Executor that is determined is actually needed, you can probably still use `can_handle` just don't have `can_handle` require a reference to `self`, in fact you can probably completely avoid the need to instantiate Executors and have all of the trait methods be static (no ref to self) just like we currently do with cgroups controllers, self is never used anyway and there's no point in instantiating then either.

Let me know what you think of this. It's possible there's an argument for dynamic dispatch that I'm missing, or maybe others prefer this approach also. Even if we decide dynamic dispatch is best I'd say the CompositeExecutor should probably just be an ExecutorManager that doesn't satisfy the Executor trait and all Executors should just be static structs that never need to be instantiated.

File: crates/libcontainer/src/workload/mod.rs [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+    /// The name of the handler
+    fn name(&self) -> &str;
+}
+pub struct CompositeExecutor {
+    executors: Vec<Box<dyn Executor>>,
+}
+
+impl Executor for CompositeExecutor {
+    fn exec(&self, spec: &Spec) -> Result<()> {
+        for executor in &self.executors {
```
[the remark this answers is not in the document]
Comment (tsturzl) [reply crates/libcontainer/src/workload/mod.rs:28]: Mind you this loop may go away if we agree against dynamic dispatch.

File: docs/src/user/webassembly.md [hunk 6]
```diff
@@ -0,0 +1,46 @@
+# Webassembly
+
+If you want to run a webassembly module with youki, your config.json has to include either **runc.oci.handler** or **module.wasm.image/variant=compat"**. 
+
+It also needs to specifiy a valid .wasm (webassembly binary) or .wat (webassembly test) module as entrypoint for the container. If a wat module is specified it will be compiled to a wasm module by youki before it is executed. The module also needs to be available in the root filesystem of the container obviously.
+
+
+```json
```
[the remark this answers is not in the document]
Comment (tsturzl) [reply docs/src/user/webassembly.md:8]: Overall, I'd say it's probably worth providing a couple of links to helpful references here. Like the links I posted in my comment, and some of the references you linked in the PR comments.

**Comment by tsturzl** [comment]
@utam0k Could you also look through my review and provide your input?

**Comment by utam0k** [comment]
@Furisto 
As this feature is very interesting, I think that this should be written in README.md. Can I ask you to write?

**Comment by utam0k** [comment]
> First of all this is really cool stuff, and it's awesome that you took the time and initiative to do this work. So thank you for that. I'm not entirely on board with using dynamic dispatch here. I think perhaps I'd consider refactoring to get rid of the `CompositeExecutor` and opt for a `ExecutorManager` that doesn't need to satisfy the `Executor` trait. This `ExecutorManager` would instead select hard coded `Executors` with a `match` block, and therefore only instantiate and allocate the `Executor` that is determined is actually needed. You can probably still use `can_handle` just don't have `can_handle` require a reference to `self`, in fact you can probably completely avoid the need to instantiate `Executors` and have all of the trait methods be static (no ref to self) just like we currently do with cgroups controllers, self is never used anyway (other than `CompositeExecutor`) and there's no point in instantiating these empty structs.
> 
> Let me know what you think of this. It's possible there's an argument for dynamic dispatch that I'm missing, or maybe others prefer that approach also. Even if we were to decide dynamic dispatch is best I'd say the `CompositeExecutor` should probably just be an `ExecutorManager` that doesn't satisfy the `Executor` trait and all `Executors` should just be static structs that never need to be instantiated.

@Furisto 
First of all, now this design is not bad at all. But I agree with @tsturzl's option because perhaps executor will only increase the number of types that can be adequately covered by hard coding. It would be better to have an ExecutorManger instead of CompositeExecutor. If you have a reason for choosing the current design, I'd like to hear it.

**Comment by utam0k** [comment]
If so, may it be difficult to separate `wasm-wasmer` feature?

**Comment by Furisto** [comment]
@tsturzl @utam0k Instead of responding to individual comments I think it is better for me to explain what my current thinking around this feature is and where I want to go with this.

Quite a few webassembly runtimes exist nowadays with the most prominent ones being wasmer, wasmtime and wasmedge. These runtimes offer different features and make different tradeoffs (wasmedge for example currently does not support compiling .wat files while wasmer does). Therefore I plan to support different webassembly runtimes to accommodate different user requirements.  

It is likely that the current method of activating a handler using a feature flag will go away as I would like to avoid that everyone has to build youki with the appropriate configuration for their specific use case. Statically linking support for all webassembly runtimes would be problematic as this would bloat the size of the youki binary considerably (even with just wasmer the size of the binary roughly doubles) and would not be of use to anyone who is not wanting to run webassembly workloads (which is still the majority for the foreseeable future). Therefore dependencies will probably linked dynamically in the future. 

I am also thinking about allowing users to to specify the preferred order in which webassembly runtimes are asked to handle the workload in the config.json. It might even be possible that a configuration file located at a well known location can be provided that allows enables/disables certain handlers globally. This might be overkill though and we would want to check if other runtimes also plan to support such a feature.

Lastly one intended use case for this will be running hooks in a webassembly sandbox, so hooks will have to be adapted as well in the future.

What I want to say with this is: I would not focus to much on how exactly an executor is called because this is likely to change anyway. I would like to get this feature in first and then iterate on it with the knowledge gained from implementing support for further webassembly runtimes.

I welcome comments on the general direction of this feature.

**Comment by utam0k** [comment]
> @tsturzl @utam0k Instead of responding to individual comments I think it is better for me to explain what my current thinking around this feature is and where I want to go with this.
> 
> Quite a few webassembly runtimes exist nowadays with the most prominent ones being wasmer, wasmtime and wasmedge. These runtimes offer different features and make different tradeoffs (wasmedge for example currently does not support compiling .wat files while wasmer does). Therefore I plan to support different webassembly runtimes to accommodate different user requirements.
> 
> It is likely that the current method of activating a handler using a feature flag will go away as I would like to avoid that everyone has to build youki with the appropriate configuration for their specific use case. Statically linking support for all webassembly runtimes would be problematic as this would bloat the size of the youki binary considerably (even with just wasmer the size of the binary roughly doubles) and would not be of use to anyone who is not wanting to run webassembly workloads (which is still the majority for the foreseeable future). Therefore dependencies will probably linked dynamically in the future.
> 
> I am also thinking about allowing users to to specify the preferred order in which webassembly runtimes are asked to handle the workload in the config.json. It might even be possible that a configuration file located at a well known location can be provided that allows enables/disables certain handlers globally. This might be overkill though and we would want to check if other runtimes also plan to support such a feature.
> 
> Lastly one intended use case for this will be running hooks in a webassembly sandbox, so hooks will have to be adapted as well in the future.
> 
> What I want to say with this is: I would not focus to much on how exactly an executor is called because this is likely to change anyway. I would like to get this feature in first and then iterate on it with the knowledge gained from implementing support for further webassembly runtimes.
> 
> I welcome comments on the general direction of this feature.

Great thoughts :100: 
I'm sorry if I'm understanding this wrong, but I think tsturzl is imagining this kind of code. I guess this should keep the byte size the same.
```rust
#[cfg(not(feature = "wasm-wasmer"))]
fn choose_executor() -> Result<Box<dyn Executor>> {
    Ok(Box::new(DefaultExecutor {}))
}

#[cfg(feature = "wasm-wasmer")]
fn choose_executor() -> Result<Box<dyn Executor>> {
    if true {
        Ok(Box::new(WasmerExecutor {}))
    } else {
        Ok(Box::new(DefaultExecutor {}))
    }
}
```

**Comment by Furisto** [comment]
@utam0k @tsturzl PTAL

**Review by utam0k — APPROVED** [review]
LGTM
Thank you for accepting our suggestion!

**Comment by Furisto** [comment]
@tsturzl seems to be busy, so I will go ahead and merge this now. But don't worry, there will be a new opportunity in the next PR for this.

**Comment by tsturzl** [comment]
Apologies for my absence. I've been on-boarding new hires at work. I did take a brief look through, and thought this looked fine. I'm excited to see the feature get merged in. I may hack on it a bit on my free time to familiarize myself with the code base.

**Pull request merged**